fix assumevalid is ignored during reindex

Eunovo opened the pull request:
During a reindex, the assumevalid setting may be ignored if the chainwork of the best header is below `minimumchainwork`. This happens when the previous IBD was stopped before connecting enough blocks to reach the required `minimumchainwork`. The assumevalid block will also be missing from the block index because the assumevalid block is set to match the `minimumchainwork`. See [Issue #31494](https://github.com/bitcoin/bitcoin/issues/31494).

As a result, `reindex` can take significantly longer to complete than necessary.

This PR addresses the issue by modifying reindex behaviour: if the chainwork of the `best_header` is below `minimumchainwork`, the node skips the `ActivateBestChain()` step and proceeds directly to synchronisation.

This PR takes a different approach from [PR #31615](https://github.com/bitcoin/bitcoin/pull/31615) to keep the validation path for `reindex` the same as the regular IBD validation path.

Developers benchmarking IBD with `-reindex` may find this behaviour undesirable. In such cases, it’s recommended to use `-reindex-chainstate` instead or ensure that the available blocks have sufficient chainwork.

`libbitcoinkernel` exposes the `ImportBlocks` function and expects `reindex` to connect blocks, so an attempt was made to preserve the existing behaviour for `libbitcoinkernel`.

Eunovo committed:
test/reindex: only connect minchainwork chain

mzumsande reviewed (commented):
Have you tried this out on signet or mainnet?
I haven't yet, but just from looking at the code I would suspect that we now 

1. do the -reindex part without connecting any blocks (except genesis [here](https://github.com/bitcoin/bitcoin/blob/138726a6f8101e0fe7e9ae701ef17b37fcbdee73/src/validation.cpp#L5161))
2. omplete headers-sync with a peer until minchainwork
3.  Download the first block we don't have on disk from a peer and only then call `ActivateBestChain()`
4. try to connect that block, which will result in the entire chain up to that block being connected by the `msghand` thread. This could block the thread hours, in which no other peer would be processed, so existing peers will get disconnected (ping timeout), while net thread will still continually make new connections (that never complete the version handshake because we are too busy connecting blocks).

Also, attempting to reindex while offline wouldn't work anymore if we are below minchainwork.

Eunovo committed:
importblocks: ensure genesis block is activated

This commit sets the stage for future commits that will change the behavior of ImportBlocks to conditionally ActivateBestChain after reindex

Eunovo committed:
importblocks: don't connect reindexed low work chains

This change will allow headers sync to discover a long enough chain before connecting blocks.
`force_activation` is required to preserve `ImportBlocks` behavior in libbitcoinkernel.

Eunovo commented:
> Also, attempting to reindex while offline wouldn't work anymore if we are below minchainwork.

I don't think it is particularly important that reindex be able to complete offline. As @hodlinator pointed out:

> Not the cleanest perhaps. But we have 2 main usages of `-reindex` AFAIK:
> 
> * Regular node-runners which run into some kind of disk corruption issue - they should typically have peers available to sync headers from.
> * Developers doing benchmarks/other work who either have sufficient block data or are okay with lowering `-minimumchainwork` temporarily.

In both use cases, network access is not an issue.

> 
> Would it be acceptable to emit an error and halt, saying that `-reindex` for cases where nodes don't have sufficient block data also requires decreasing `-minimumchainwork=` to whatever level we have in the block data? Could also explain that this is in order to keep logic consistent regardless of `-reindex`, and also say that running with lowered `-minimumchainwork` long-term is not advisable.

It is possible to do this, but will this be considered a better solution than automatically doing headers sync?

> I haven't checked the code yet (am planning on doing that later), but would it be possible to just issue a warning if we're offline and attempt the reindex anyway (especially now that we have proper script verification logging to show if script verification was enabled)?

IBD does not work offline anyway. I'm not sure this will be better to implement, unless there is a use-case that requires `-reindex` to work offline with a low-work chain.

Eunovo committed:
notifications: add header_tip mutex

The background init thread will wait on this mutex for header tip updates to determin whether to ActivateBestChain  in future commits..

Eunovo committed:
net: allow headers sync during reindex

This commit removes:
- LoadingBlocks() guard that prevent HEADERS sync during reindex, and
- adds a LoadingBlocks() guard that prevents requesting blocks during reindex

Blocks must not be requested while LoadingBlocks() is `true` because they will be
rejected when received and this might lead to the peer being disconnected because
the requested block will never be removed from vBlocksInFlight.

Note that we cannot connect Blocks while LoadingBlocks() is `true` because this can
block the p2p process message thread for a long time if a long chain is being connected
in the background.

Eunovo committed:
test: update anti-dos test

Future commits will change the LoadingBlocks stage to only end after best_header has at least minchainwork.
Mining a block at the tip will no longer be enough to trigger the "unrequested block" scenario.

Eunovo committed:
init: keep bg init thread until headers sync is done

LoadingBlocks() will now remain true until headers sync is complete.
To preserve `btck_chainstate_manager_import_blocks` behaviour,
`ImportBlocks` will set `m_blockman.m_importing` if called with
`force_activation = True`, otherwise, `ImportBlocks` has been called
from the background init thread and we assert that `m_blockman.m_importing`
is already set.

Eunovo commented:
> This suggestion was meant for the case where we can't do headers sync due to being offline (using zero peers after a certain timeout or something like that, I'm not sure how to best check whether we are offline). I guess the error message could also point out that going online should resolve the situation.

I see. A message that indicates that the node is waiting for headers sync to finish reindex will be valuable. We probably don't need to check that we are offline at all.

Eunovo committed:
test: script ver is skipped during reindex with -assumevalid

Test that `-assumevalid` is not ignored even when `minchainwork` is higher
than the `best_header` on disk during reindex.

Eunovo commented:
Thanks for the reviews.

@l0rinc I added your test as an automated test in https://github.com/bitcoin/bitcoin/pull/33854/commits/729bf4d627f3bc3414af22990fcf67af2d50d4cf
@hodlinator I added a useful log message in https://github.com/bitcoin/bitcoin/pull/33854/commits/729bf4d627f3bc3414af22990fcf67af2d50d4cf
